commit 7bb1efdc05f048ce49fd3209048d76a2e559de14
Author: Wolfgang Rohdewald <wolfgang@rohdewald.de>
Date:   Fri Oct 3 23:28:27 2014 +0200

    Ticket 7661: twisted.spread.banana._selectDialect: selecting an unsupported dialect now raises BananaError
    
    This may happen if the client process asks the server to use an unsupported dialect.

diff --git twisted/spread/banana.py twisted/spread/banana.py
index 82b7256..0b215eb 100644
--- twisted/spread/banana.py
+++ twisted/spread/banana.py
@@ -110,6 +110,8 @@ class Banana(protocol.Protocol, styles.Ephemeral):
         """
 
     def _selectDialect(self, dialect):
+        if dialect not in Banana.knownDialects:
+            raise BananaError("%r is an unknown dialect" % (dialect,))
         self.currentDialect = dialect
         self.connectionReady()
 
diff --git twisted/spread/topfiles/7661.bugfix twisted/spread/topfiles/7661.bugfix
new file mode 100644
index 0000000..57f5afa
--- /dev/null
+++ twisted/spread/topfiles/7661.bugfix
@@ -0,0 +1 @@
+twisted.spread.banana._selectDialect now raises BananaError when selecting an unsupported dialect
diff --git twisted/test/test_banana.py twisted/test/test_banana.py
index c4b69de..d914f47 100644
--- twisted/test/test_banana.py
+++ twisted/test/test_banana.py
@@ -45,6 +45,16 @@ class BananaTestCase(unittest.TestCase):
         self.enc.dataReceived(self.io.getvalue())
         assert self.result == 'hello'
 
+
+    def test_validDialect(self):
+        """
+        banana only knows dialects none and pb
+        """
+        self.enc._selectDialect(b'none')
+        self.enc._selectDialect(b'pb')
+        self.assertRaises(banana.BananaError, self.enc._selectDialect, b'abc')
+
+
     def test_int(self):
         """
         A positive integer less than 2 ** 32 should round-trip through
